chia/full_node/full_node.py

```diff
@@ -1111,27 +1112,75 @@ async def sync_from_fork_point(
+            # be sent. It's initialized to the current time, and bumped by the
+            # seconds_per_request every time we send a request. This ensures we
+            # won't exceed the 100 requests / 60 seconds rate limit.
+            # Whichever peer has the lowest timestamp is the one we request
+            # from. peers that take more than 5 seconds to respond are pushed to
+            # the end of the queue, to be less likely to request from.
+
+            # This should be cleaned up to not be a hard coded value, and maybe
+            # allow higher request rates (and align the request_blocks and
+            # respond_blocks rate limits).
+            now = time.monotonic()
+            new_peers_with_peak: list[tuple[WSChiaConnection, float]] = [(c, now) for c in peers_with_peak[:]]
+            self.log.info(f"peers with peak: {len(new_peers_with_peak)}")
+            random.shuffle(new_peers_with_peak)
             try:
                 # block request ranges are *inclusive*, this requires some
                 # gymnastics of this range (+1 to make it exclusive, like normal
                 # ranges) and then -1 when forming the request message
                 for start_height in range(fork_point_height, target_peak_sb_height + 1, batch_size):
                     end_height = min(target_peak_sb_height, start_height + batch_size - 1)
                     request = RequestBlocks(uint32(start_height), uint32(end_height), True)
+                    new_peers_with_peak.sort(key=lambda pair: pair[1])
                     fetched = False
-                    for peer in random.sample(new_peers_with_peak, len(new_peers_with_peak)):
+                    for idx, (peer, timestamp) in enumerate(new_peers_with_peak):
                         if peer.closed:
                             continue
+
                         start = time.monotonic()
+                        if start < timestamp:
+                            # rate limit ourselves, since we sent a message to
+                            # this peer too recently
+                            await asyncio.sleep(timestamp - start)
+                            start = time.monotonic()
+
+                        # update the timestamp, now that we're sending a request
+                        if is_localhost(peer.peer_info.host):
+                            # we don't apply rate limits to localhost, and our
+                            # tests depend on it
+                            new_peers_with_peak[idx] = (
+                                new_peers_with_peak[idx][0],
+                                new_peers_with_peak[idx][1] + 0.1,
+                            )
+                        else:
+                            new_peers_with_peak[idx] = (
+                                new_peers_with_peak[idx][0],
+                                new_peers_with_peak[idx][1] + seconds_per_request,
```

Comment:
just a suggestion  i think that rather then setting the would be time for the next request maybe just keeping track of the time the last request was sent and then using 
if now > sent_time + seconds_per_request 
in the condition would be clearer

Reply:
you can add a case for local host, i think that when we update the times inside the array its harder to debug/understand whats going on rather then just seeing the original time and the delta condition in the case we go into

Reply:
the other reason to specify the time when we're allowed to send the next message, which I forgot to mention earlier, is that it makes it simpler to "deprioritize" a peer. Right now, if a peer takes too long, we clear out any accrued space to send requests, by resetting the timer to "now". This is the logic you commented on here:

https://github.com/Chia-Network/chia-blockchain/pull/18729/files#r1822155568

That logic would be a bit more complicated as we would need to subtract the pace (`seconds_per_block`) which depends on whether it's a localhost peer or not.

The key place this value is being used, is here:
https://github.com/Chia-Network/chia-blockchain/pull/18729/files#diff-a84a0802b416c9c5a5fdc6d2293c4c7971ca0b1519f246ede123e304a56e540aR1151

```
if start < timestamp:
```

Which I also think benefits from having the timestamp be the next available time. Note that `timestamp` can be in the past, if we haven't been able to keep up, peers accrue space to send more requests by these timestamps falling behind real-time.